### PR TITLE
Added media collection feature

### PR DESCRIPTION
**Pull request by s-weigand**
First of all, thanks for this great tool. 😄 
Sadly the images relative to the notebook couldn't be rendered, since ``nbsphinx``, does look for them relative to the ``. nblink`` file.

Which is why I added a media collection feature, with the "extra-media" key.This will copy all specified media files/directories, relative to the ``. nblink`` file, as they were to the notebook, which in turn leads to ``nbsphinx`` rendering the notebook properly.
If a given path is wrong, it will throw a ``sphinx warning`` telling the user that the path was wrong and in which file it was defined.

P.S.: The diff is a bit bigger than what I added, since the code was formatted with [``black``](https://github.com/psf/black).

### CODE REVIEW

**Comment by vidartf** [comment]
Thanks.

> P.S.: The diff is a bit bigger than what I added, since the code was formatted with black.

I strongly prefer to keep formatting and code changing commits separately. Could you remove the format changing code from the PR, or at thee very least ensure that it is a separate commit?

**Comment by s-weigand** [comment]
I reverted the formating changes.

File: nbsphinx_link/__init__.py [hunk 1]
```diff
@@ -11,27 +11,95 @@
...
 import nbformat
+from sphinx.util.logging import getLogger
 
 from ._version import __version__
 
 
+def copy_directory(src, dest):
+    """
+    Copies a directory or file from the path ``src`` to ``dest``.
+    Originally from: https://www.pythoncentral.io/how-to-recursively-copy-a-directory-folder-in-python/
```
Comment (vidartf) [inline nbsphinx_link/__init__.py]: pythoncentral licenses the code under the MIT license, so I think you will have to include that license text with copyright notice.

File: nbsphinx_link/__init__.py [hunk 2]
```diff
@@ -11,27 +11,95 @@
     "path": "relative/path/to/notebook"
 }
 
+Optionally the "extra-media" key can be added, if you notebook includes
+any media, i.e. images. The value can be an array or a string, with
+path/-s to the media file/-s or directory/-s.
```
Comment (vidartf) [inline nbsphinx_link/__init__.py]: Let's just have it as an array of string only (though still optional).

File: nbsphinx_link/__init__.py [hunk 3]
```diff
@@ -11,27 +11,95 @@
...
+        if e.errno == errno.ENOTDIR:
+            shutil.copy(src, dest)
+        else:
+            print('Directory not copied. Error: {}'.format(e))
+
+
+def collect_extra_media(extra_media, source_file, nb_path):
+    """
+    Collects extra media defined in the .nblink file, with the key 'extra-media'.
+    The extra media (i.e. images) need to be copied on order nbsphinx to properly
```
Comment (vidartf) [inline nbsphinx_link/__init__.py]:
```suggestion
    The extra media (i.e. images) need to be copied in order for nbsphinx to properly
```

File: nbsphinx_link/__init__.py [hunk 4]
```diff
@@ -11,27 +11,95 @@
...
+    for extract_media_path in extra_media:
+        if os.path.isabs(extract_media_path):
+            src_path = extract_media_path
+        else:
+            extract_media_relpath = os.path.join(source_dir, extract_media_path)
+            src_path = os.path.normpath(os.path.join(source_dir, extract_media_relpath))
+        target_path = utils.relative_path(nb_path, src_path)
+        target_path = os.path.normpath(os.path.join(source_dir, target_path))
+        if os.path.exists(src_path):
+            copy_directory(src_path, target_path)
```
Comment (vidartf) [inline nbsphinx_link/__init__.py]: This `copy_directory` call will blow out the contents of `target_path`, which I don't think is safe. I think this can particularly be an issue if the extra-media paths point to a folder above the notebook folder (e.g. `../../`). I think we should:
- Ensure that the destination path is not outside the docs source dir.
- Individually copy each file (overwriting if necessary), rather than an indiscriminate directory deletion.

File: .gitignore [hunk 5]
```diff
@@ -137,3 +137,10 @@ $RECYCLE.BIN/
 
 # Windows shortcuts
 *.lnk
+
+# vscode config
+.vscode
+
+# collected media
+docs/source/python_logo.png
+docs/source/images/python_logo.png
```
Comment (vidartf) [inline .gitignore]: It's a pretty clear disadvantage to this functionality that the images are simply copied verbatim into the docs source tree. I'm ok with that for now though, as it will only happen if the author explicitly requests it.

File: docs/source/introduction.nblink [hunk 6]
```diff
@@ -1,3 +1,8 @@
 {
-  "path": "../../notebooks/introduction.ipynb"
+  "path": "../../notebooks/introduction.ipynb",
+  "extra-media": [
+    "../../notebooks/images",
+    "../../notebooks/python_logo.png",
```
Comment (vidartf) [inline docs/source/introduction.nblink]: Note that this image is licensed under CC-BY-SA, which means we need to include the license and attribution (both in the source, and in the final docs). I would very much prefer to use a public domain (e.g. CC0) image, or one we create our selves.

File: nbsphinx_link/__init__.py [hunk 7]
```diff
@@ -59,6 +131,11 @@ def parse(self, inputstring, document):
         path = utils.relative_path(None, abs_path)
         path = nodes.reprunicode(path)
 
+        extra_media = link.get('extra-media', None)
+        if extra_media:
+            source_file = env.doc2path(env.docname)
+            collect_extra_media(extra_media, source_file, path)
```
Comment (vidartf) [inline nbsphinx_link/__init__.py]: I think we need to mark all our input files as a dependency (see below). This will help ensure that Sphinx re-runs this parser if some of those files change (rather than skipping this build step since the linkfile + notebook are unchanged).

**Review by vidartf — CHANGES_REQUESTED** [review]
See a detailed review below. Most important are the license issues.

File: nbsphinx_link/__init__.py (hunk 4 above)
Comment (s-weigand) [reply nbsphinx_link/__init__.py]: This should only happen, if the ``.nblink`` file would be outside of the docs folder, in which case it won't be discovered by sphinx. So I think we are good.

File: nbsphinx_link/__init__.py (hunk 1 above)
Comment (s-weigand) [reply nbsphinx_link/__init__.py]: Thanks for the hint, I will keep things like that  in my mind in the future.
I put a link to the MIT License in the doc-string as stated on the [CC BY-NC 3.0 website](https://creativecommons.org/licenses/by-nc/3.0/deed.en_US).
I hope this is fine. 
IMO copying the whole text would clutter the doc string.

File: .gitignore (hunk 5 above)
Comment (s-weigand) [reply .gitignore]: It is a bit messy, but up to the user if he/she wants to also upload the redundant copy.
In my project I have defined structure and just added all folder called ``image`` inside of the docs folder to the ``.gitignore`` (``docs/**/**/images/``).

File: docs/source/introduction.nblink (hunk 6 above)
Comment (s-weigand) [reply docs/source/introduction.nblink]: I replaced it with a smiley I did draw myself.
Thanks again for the pointer, I'm not too firm with the legal things and thought it is fine as long as you give props to the creator.

File: nbsphinx_link/__init__.py (hunk 4 above)
Comment (s-weigand) [reply nbsphinx_link/__init__.py]: I could also rewrite ``copy_directory`` to just use ``shutil.copy``, ``os.makedirs`` and ``os.walk`` if the ``src`` is a directory, which would also save the hustle with license linking of pythoncentral.

File: nbsphinx_link/__init__.py (hunk 7 above)
Comment (s-weigand) [reply nbsphinx_link/__init__.py]: I implemented it, but I'm not sure if it works properly directorys.
I tryed to trace back the definitions ``note_dependency`` and ``record_dependencies.add``, but couldn't find them.
Maybe the [mentioned rewrite of ``copy_directory ``](https://github.com/vidartf/nbsphinx-link/pull/10#discussion_r328603730), could solve this.

Comment (vidartf) [reply nbsphinx_link/__init__.py]: I think we can use [`note_reread`](https://www.sphinx-doc.org/en/master/extdev/envapi.html#sphinx.environment.BuildEnvironment.note_reread):
- If `extra-media` includes a directory, call `note_reread` to make sure the current file is always rebuilt. Each individual file should still be added to `record_dependencies` (and probably also good to add with `note_dependency`).
- If `extra-media` is not present, or only has files, add all the files with `note_dependency` etc.

File: README.rst [hunk 8]
```diff
@@ -24,6 +24,10 @@ The .nblink file is a JSON file with the following structure::
         "path": "relative/path/to/notebook"
     }
 
+Optionally the "extra-media" key can be added, if you notebook includes
+any media, i.e. images. The value can be an array or a string, with
+path/-s to the media file/-s or directory/-s.
```
Comment (vidartf) [inline README.rst]:
```suggestion
paths to the media files or directories.
```

File: docs/source/conf.py [hunk 9]
```diff
@@ -117,7 +117,7 @@
 # Add any paths that contain custom static files (such as style sheets) here,
 # relative to this directory. They are copied after the builtin static files,
 # so a file named "default.css" will overwrite the builtin "default.css".
-html_static_path = ['_static']
+# html_static_path = ['_static']
```
Comment (vidartf) [inline docs/source/conf.py]: Is this just cleanup, or is this change necessary for the code to work?

File: nbsphinx_link/__init__.py (hunk 1 above)
Comment (vidartf) [reply nbsphinx_link/__init__.py]: The licensing term requires the entire text to be included (the CC guidelines do not affect the MIT license), but it doesn't need to be in the docstring, it can e.g. be in a comment.

File: nbsphinx_link/__init__.py (hunk 7 above)
Comment (s-weigand) [reply nbsphinx_link/__init__.py]: Wouldn't ``note_reread`` rerender the notebook no matter if it changed or not?

File: docs/source/conf.py (hunk 9 above)
Comment (s-weigand) [reply docs/source/conf.py]: ``sphinx`` just trows a warning that the folder ``_static`` does't exist if ``html_static_path`` is set.

File: nbsphinx_link/__init__.py (hunk 7 above)
Comment (vidartf) [reply nbsphinx_link/__init__.py]: > Wouldn't `note_reread` rerender the notebook no matter if it changed or not?

Yes, but that is the only way to ensure that all changes in a media directory gets included. Take this case:
1. You add a reference to an image 'images/bob.png' in a markdown cell, add the dir to the media field, and then make the docs, which see the changes to the notebook and rebuilds. ✔️ 
2. You want to replace the image with another, so you delete 'bob.png', but forget to copy over the new image (or the copy op fails for some reason). You make the docs, which see the changes to bob.png (deletion) and rebuilds. ✔️
3. You see the broken image in your docs, and realize your mistake. You add the new image and make the docs. No previously used file (link, notebook, or image file) has changed, so it skips the build. ❌

Comment (vidartf) [reply nbsphinx_link/__init__.py]: (Actually, the above example might fail at step 2 due to a missing reference, and failed build will always rebuild, but the concept holds for other examples)

File: nbsphinx_link/__init__.py (hunk 4 above)
Comment (vidartf) [reply nbsphinx_link/__init__.py]: > I could also rewrite copy_directory to just use shutil.copy, os.makedirs and os.walk if the src is a directory, which would also save the hustle with license linking of pythoncentral.

I think this is the right solution 👍

File: docs/source/conf.py (hunk 9 above)
Comment (vidartf) [reply docs/source/conf.py]: So clean-up then. Thanks!

**Comment by s-weigand** [comment]
I rewrote the copy function to just copy files (no more deletion) and also register each file seperately as a dependency.

File: nbsphinx_link/__init__.py (hunk 7 above)
Comment (s-weigand) [reply nbsphinx_link/__init__.py]: Now each file is registered indiviualy with ``note_dependency`` and ``record_dependencies``

File: nbsphinx_link/__init__.py [hunk 10]
```diff
@@ -19,53 +19,91 @@
...
+    Copies a singe file from ``src`` to ``dest``.
+
+    Parameters
+    ----------
+    src : str
+        Path to the source file.
+    dest : str
+        Path to the destination file.
+    """
+    logger = getLogger(__name__)
```
Comment (vidartf) [inline nbsphinx_link/__init__.py:65]: We should use the Sphinx logger instead.

File: nbsphinx_link/__init__.py [hunk 11]
```diff
@@ -19,53 +19,91 @@
...
-        if e.errno == errno.ENOTDIR:
-            shutil.copy(src, dest)
-        else:
-            print('Directory not copied. Error: {}'.format(e))
+    if os.path.isdir(src):
+        for root, _, filenames in os.walk(src):
+            for filename in filenames:
+                src_path = os.path.abspath(os.path.join(root, filename))
+                if not os.path.exists(dest):
+                    os.makedirs(dest)
```
Comment (vidartf) [inline nbsphinx_link/__init__.py]: This existence check seems like it does not belong here. It should be moved outside the filename loop, and should check for and create `os.path.join(dest, os.relpath(root, src))` (I think, probably needs tweaking). You could also just call os.makedirs with `exist_ok=True`.

**Review by vidartf — COMMENTED** [review]
Getting closer :)

File: nbsphinx_link/__init__.py (hunk 7 above)
Comment (vidartf) [reply nbsphinx_link/__init__.py]: But no `note_reread`?

File: nbsphinx_link/__init__.py (hunk 10 above)
Comment (vidartf) [reply nbsphinx_link/__init__.py:65]: I.e. `sphinx.util.logging.getLogger(__name__)`, not the deprecated logging API.

Comment (s-weigand) [reply nbsphinx_link/__init__.py:65]: Its is ``from sphinx.util.logging import getLogger`` or do you have a different logger in mind?

File: nbsphinx_link/__init__.py (hunk 7 above)
Comment (s-weigand) [reply nbsphinx_link/__init__.py]: Maybe I missunderstood your comment:

> If extra-media is not present, or only has files, add all the files with note_dependency etc.

File: nbsphinx_link/__init__.py (hunk 10 above)
Comment (vidartf) [reply nbsphinx_link/__init__.py:65]: I already commented here, but I think you might have missed the second comment? :)

File: nbsphinx_link/__init__.py (hunk 11 above)
Comment (s-weigand) [reply nbsphinx_link/__init__.py]: I moved it out if the loop.
``exist_ok=True`` would break the py2.7 compatibility, since it was [added in python 3.2](https://docs.python.org/3.7/library/os.html#os.makedirs) .

File: nbsphinx_link/__init__.py (hunk 10 above)
Comment (s-weigand) [reply nbsphinx_link/__init__.py:65]: Looks like we talk about the same ``logger``, so it should be fine. :)

File: nbsphinx_link/__init__.py (hunk 7 above)
Comment (s-weigand) [reply nbsphinx_link/__init__.py]: Do you want me add ``note_reread`` on a per file basis, for the extra media files?
I would add a ``is_extra_media`` kwarg to ``register_dependency``.

File: nbsphinx_link/__init__.py (hunk 11 above)
Comment (vidartf) [reply nbsphinx_link/__init__.py]: Ok on the py2 bit. I pushed a commit to ensure that this code handles "src" having subdirectories. Do you agree with the changes?

File: nbsphinx_link/__init__.py (hunk 7 above)
Comment (vidartf) [reply nbsphinx_link/__init__.py]: I meant: If `extra-media` is present, and contains at least one directory, then run `document.settings.env.note_reread()` once. The other code should run in all cases (might not be needed, but let's be defensive).

File: nbsphinx_link/__init__.py (hunk 11 above)
Comment (s-weigand) [reply nbsphinx_link/__init__.py]: Thats nice, i will also add an example for that in the notebook.

Comment (s-weigand) [reply nbsphinx_link/__init__.py]: I made ``dst_root`` also the target to copy files to.

File: nbsphinx_link/__init__.py (hunk 7 above)
Comment (s-weigand) [reply nbsphinx_link/__init__.py]: Now it runs once if ``extra-media`` is present at all.
Else we would need an extra loop/map, after copying the files, to check if any item of ``extra_media`` is a directory, so it won't run on each directory.

Comment (vidartf) [reply nbsphinx_link/__init__.py]: I pushed a commit that checks each config arg as part of the outer loop, and call `note_reread` the first time a dir is found.

Comment (vidartf) [reply nbsphinx_link/__init__.py]: Let me know if you agree with the change.

File: nbsphinx_link/__init__.py [hunk 12]
```diff
@@ -58,7 +58,7 @@ def copy_file(src, dest, document):
     src : str
         Path to the source file.
     dest : str
-        Path to the destination file.
+        Path to the destination file or directory.
```
Comment (s-weigand) [inline nbsphinx_link/__init__.py:61]: ``copy_file`` does only copy a single file, since ``shutil.copy`` does only copy a single file.

File: nbsphinx_link/__init__.py [hunk 13]
```diff
@@ -140,6 +141,9 @@ def collect_extra_media(extra_media, source_file, nb_path, document):
         dest_path = utils.relative_path(nb_path, src_path)
         dest_path = os.path.normpath(os.path.join(source_dir, dest_path))
         if os.path.exists(src_path):
+            if not any_dirs and os.path.isdir(src_path):
+                any_dirs = True
+                document.settings.env.note_reread()
```
Comment (s-weigand) [inline nbsphinx_link/__init__.py]: I think ``document.settings.env.note_reread()`` should be called at the end of ``collect_extra_media``.

```python
if  any_dirs:
    document.settings.env.note_reread()
```
Else ``any_dirs `` is pointless.

File: nbsphinx_link/__init__.py [hunk 14]
```diff
@@ -11,27 +11,155 @@
...
 """
 
-
 import json
 import os
+import shutil
 
 from docutils import io, nodes, utils
 from docutils.utils.error_reporting import SafeString, ErrorString
+import docutils  # noqa: F401
```
Comment (vidartf) [inline nbsphinx_link/__init__.py:28]: ?

**Review by vidartf — APPROVED** [review]

File: nbsphinx_link/__init__.py (hunk 13 above)
Comment (vidartf) [reply nbsphinx_link/__init__.py]: `any_dirs` is only there to shortcut the if condition (i.e. prevent `isdir` from being called). Maybe the simpler version reads better, and the performance gain is negligible. I'll change it.

File: nbsphinx_link/__init__.py (hunk 14 above)
Comment (s-weigand) [reply nbsphinx_link/__init__.py:28]: The ``# noqa: F401`` if for ``flake8`` not to complain about the unused import, since ``docutils`` is only used in the docsting for the type of ``document``.

File: nbsphinx_link/__init__.py (hunk 13 above)
Comment (vidartf) [reply nbsphinx_link/__init__.py]: Pushed

Comment (s-weigand) [reply nbsphinx_link/__init__.py]: Ups my bad, never mind (should get another coffee) >.<

File: nbsphinx_link/__init__.py (hunk 12 above)
Comment (vidartf) [reply nbsphinx_link/__init__.py:61]: Yes, but the *destination* can be either a full file path, or a directory path.

Comment (s-weigand) [reply nbsphinx_link/__init__.py:61]: Oh yes, true that.
